core/lib/zksync_core/src/api_server/web3/pubsub.rs

```diff
@@ -238,118 +238,174 @@ impl EthSubscribe {
+        let sink = pending_sink.accept().await.unwrap();
+        let closed = sink.closed().fuse();
+        let mut b = BroadcastStream::new(b);
+        tokio::pin!(closed);
+
+        loop {
+            tokio::select! {
+                new_log = b.next() => {
+                    if let Some(Ok(new_log)) = new_log {
+                        sink.send_timeout(
+                            SubscriptionMessage::from_json(&PubSubResult::Header(new_log))
+                            .expect("PubSubResult always serializable to json; qed"),
+                            SUBSCRIPTION_SINK_SEND_TIMEOUT,
+                        )
+                            .await.ok();
+                        } else {
+                            break
+                    }
+                }
+                _ = &mut closed => {
+                    break
+                }
+            }
+        }
+    }
+
+    async fn transactions_subscriber(
+        pending_sink: PendingSubscriptionSink,
+        b: broadcast::Receiver<H256>,
+    ) {
+        PUB_SUB_METRICS.active_subscribers[&SubscriptionType::Txs].inc_guard(1);
+        let sink = pending_sink.accept().await.unwrap();
+        let closed = sink.closed().fuse();
+        let mut b = BroadcastStream::new(b);
+        tokio::pin!(closed);
+
+        loop {
+            tokio::select! {
+                new_log = b.next() => {
+                    if let Some(Ok(new_log)) = new_log {
+                        sink.send_timeout(
+                            SubscriptionMessage::from_json(&PubSubResult::TxHash(new_log))
+                            .expect("PubSubResult always serializable to json; qed"),
+                            SUBSCRIPTION_SINK_SEND_TIMEOUT,
+                        )
+                            .await.ok();
+                        } else {
+                            break
+                    }
+                }
+                _ = &mut closed => {
+                    break
+                }
+            }
+        }
     }
 
-    fn reject(subscriber: typed::Subscriber<PubSubResult>) {
-        subscriber
-            .reject(Error {
-                code: ErrorCode::InvalidParams,
-                message: "Rejecting subscription - invalid parameters provided.".into(),
-                data: None,
-            })
-            .unwrap();
+    async fn logs_subscriber(
+        pending_sink: PendingSubscriptionSink,
+        b: broadcast::Receiver<Log>,
+        filter: PubSubFilter,
+    ) {
+        let _guard = PUB_SUB_METRICS.active_subscribers[&SubscriptionType::Logs].inc_guard(1);
+        let sink = pending_sink.accept().await.unwrap();
+        let closed = sink.closed().fuse();
+        let mut b = BroadcastStream::new(b);
+        tokio::pin!(closed);
+
+        loop {
+            tokio::select! {
+                new_log = b.next() => {
+                    let new_log = new_log.unwrap().unwrap();
```

Comment:
I don't think `unwrap()`s are safe here; we may get `None` if the notifier task terminates, and an error if `broadcast::Receiver` is too far behind.

Reply:
removed